CASSGO-49 fix for the cassandra version unmarshal

tengu-alt opened the pull request:
Fix for the #910. 
This PR fixes issue, when the driver is unable to unmarshal Cassandra version, which contains additional annotation.

testisnullus commented on host_source.go:
```diff
@@ -56,6 +56,7 @@ const (
 
 type cassVersion struct {
 	Major, Minor, Patch int
+	AdditionalNotation  string
```
In Cassandra (and many Java-based projects), the bit after the dash is often called a qualifier or sometimes a suffix. In semantic versioning, it might be part of a pre-release segment or a build meta after a plus sign. However, Cassandra uses variations like snapshot, rc, etc., which are typically called qualifiers.

So, I suggest changing the name of the AdditionalNotation field to the Qualifier.

joao-r-reis replied:
+1 to this suggestion

joao-r-reis commented on host_source.go:
```diff
@@ -87,13 +88,29 @@ func (c *cassVersion) unmarshal(data []byte) error {
+		if err != nil {
+			return fmt.Errorf("invalid minor version %v: %v", v[1], err)
+		}
+		c.AdditionalNotation = vMinor[1]
```
Should return here as we're parsing the qualifier which should always be at the end. 

Also if the "suffix" of the version is something like `-qualifier1-qualifier2` we want this field to either contain the full suffix string `-qualifier1-qualifier2` or this field should be of type `[]string` and the slice should be `[qualifier1,qualifier2]`. [Here's the relevant code in the java driver](https://github.com/apache/cassandra-java-driver/blob/4.x/core/src/main/java/com/datastax/oss/driver/api/core/Version.java#L122-L126). 

Personally I think it's fine if we just parse the full suffix string including the first `-`, if anyone is actually interested in parsing these qualifiers then they can do it on their app. In this case the name of the field should probably be `Suffix` instead of `Qualifier` though I think?

jameshartig replied:
How common will there be multiple qualifiers? I'm fine with `Suffix` but if there isn't realistically going to be more than one qualifier, then I think `Qualifier` is still more appropriate.

joao-r-reis replied:
Honestly I don't know, I just noticed there's logic to handle this in the java driver and thought we would do the same here but I don't remember the last time I saw a C* version with multiple qualifiers.

joao-r-reis reviewed (commented):
Added some comments. Let's get this in gocql 2.0, can you create a JIRA?

tengu-alt commented:
@joao-r-reis , @testisnullus I have renamed the `AdditionalNotation` field to the `Suffix`, and refactored the way we parse it in case it has more than one `qualifier`. I decided to retrieve it as one string. 

> Personally I think it's fine if we just parse the full suffix string including the first `-`

Could you please clarify why we need to contain the first `-`? I don't get the point of it.

jameshartig commented:
I'm fine with including the `-` if its being called `Suffix` but I also don't know if its a big deal either way. It does look like the java driver removes the leading dashes but it also exposes an array.

joao-r-reis commented:
James suggested we should keep `Qualifier` because we don't see versions with multiple qualifiers often (or at all recently). I think we can keep the first `-` out of it as it is currently on this PR and maybe just renaming it `Qualifier`, are you ok with this @jameshartig ?

jameshartig commented:
> James suggested we should keep `Qualifier` because we don't see versions with multiple qualifiers often (or at all recently). I think we can keep the first `-` out of it as it is currently on this PR and maybe just renaming it `Qualifier`, are you ok with this @jameshartig ?

Yep! That sounds good to me

joao-r-reis commented:
👍 @tengu-alt all we need is to rename `Suffix` to `Qualifier` and we're good to go then

tengu-alt committed:
Cassandra version unmarshal fix

FIx for the issue, when the driver is unable to unmarshal Cassandra version,
which contains additional annotation (suffix).

patch by Oleksandr Luzhniy; reviewed by João Reis, James Hartig, Danylo Savchenko, for CASSGO-49

tengu-alt commented:
Done, also updated the commit message and the changelog.

tengu-alt commented:
@joao-r-reis should it be merged?
